File: shared/js/src/services/healthcheck/healthcheck.ts
import { Model } from '../../types/model'
export enum HealthCheckStatus {
  HEALTHY = 'healthy',
  UNHEALTHY = 'unhealthy',
  DEGRADED = 'degraded'
}

export type HealthCheckExecutorType = 'default' | string

export type HealthCheckResponse = {
  status: HealthCheckStatus
}

export interface HealthCheck extends Model {
  name: string
  description?: string
  url: string
  interval: number
  active: boolean
  method?: string
  expectedResponseCode?: number
  timeout?: number
  auth?: {
    username: string
    password: string
    apiKey: string
  }
  executor?: HealthCheckExecutorType
}

export const withDefaults = (healthCheck: HealthCheck): HealthCheck => {
  return {
    id: healthCheck.id,
    name: healthCheck.name,
    description: healthCheck.description || '',
    url: healthCheck.url,
    active: healthCheck.active,
    interval: healthCheck.interval || 30000,
    method: healthCheck.method || 'GET',
    expectedResponseCode: healthCheck.expectedResponseCode || 200,
    timeout: healthCheck.timeout || 5000,
    auth: healthCheck.auth,
    executor: healthCheck.executor || 'default'
  }
}

File: go/internal/models/healthcheck.go
package models

type HealthCheckExecutorType string

const (
	CURL    HealthCheckExecutorType = "curl"
	DEFAULT HealthCheckExecutorType = "default"
)

type HealthCheck struct {
	Id                   string                  `json:"id"`
	Name                 string                  `json:"name"`
	Description          string                  `json:"description,omitempty"`
	URL                  string                  `json:"url"`
	Interval             int                     `json:"interval"`
	Active               bool                    `json:"active"`
	Method               string                  `json:"method,omitempty"`
	ExpectedResponseCode int                     `json:"expectedResponseCode,omitempty"`
	Timeout              int                     `json:"timeout,omitempty"`
	Executor             HealthCheckExecutorType `json:"executor,omitempty"`
	Auth                 *struct {
		Username string `json:"username,omitempty"`
		Password string `json:"password,omitempty"`
		APIKey   string `json:"apiKey,omitempty"`
	} `json:"auth,omitempty"`
}
